dynawo/sources/ModelicaCompiler/Scripts_OMC_1_13_2/factory.py

```diff
@@ -512,52 +522,195 @@ def build_call_functions(self):
-        num_body = 0
-        for body in list_body_to_analyse:
-            for name in name_func_to_search:
-                if any(name in line for line in body):
-                    body_to_add[num_body] = True
-                    break
 
-            num_body += 1
+        global_pattern_index = 0
+        dic_var_name_to_temporary_name = {}
 
-        # Add the functions concerned
-        num_body = 0
+        # functions calling an external function ???
         for body in list_body_to_analyse:
-            if body_to_add[num_body]:
-                self.list_call_for_setf.append('  {\n')
+            new_body = []
+            is_discrete = False
+            found = False
+            for line in body:
+                for name in name_func_to_search:
+                    if name+"(" in line or name +" (" in line:
+                        for func in filter(lambda x: (x.get_name() == name), list_omc_functions):
+                            variables_set_by_omc_function = self.find_variables_set_by_omc_function(line, func, None)
+                            discrete_variables_set_by_omc_function = self.find_output_integer_double_vars(line, func)
+
+                            ## Sanity check: cannot assign a value to both a continuous and a discrete variable
+                            if len(discrete_variables_set_by_omc_function) > 0 and len(discrete_variables_set_by_omc_function) < len(variables_set_by_omc_function):
+                                error_exit("    Error: Function " + name + " is used to assign a value to both continuous and discrete variables, which is forbidden.")
+
+                            if len(discrete_variables_set_by_omc_function) > 0:
+                                is_discrete = True
+                            else:
+                                (line, dic_var_name_to_temporary_name_tmp, global_pattern_index) = self.replace_var_names_by_temporary_and_build_dictionary(line, variables_set_by_omc_function, global_pattern_index)
+                                dic_var_name_to_temporary_name.update(dic_var_name_to_temporary_name_tmp)
+                        found = True
+                if "omc_assert" in line or "omc_terminate" in line:
+                    found = True
+                new_body.append(line)
+            if found and is_discrete:
+                list_body_to_append.append("{\n")
+                list_body_to_append_to_z.append(body)
+                list_body_to_append.append("}\n\n")
+            elif found:
+                list_body_to_append.append("{\n")
+                list_body_to_append.append(new_body)
+                list_body_to_append.append("}\n\n")
+
+        if len(list_body_to_append) > 0:
+            for name in dic_var_name_to_temporary_name.keys():
+                test_param_address(name)
+                self.list_call_for_setf.append("  double " + dic_var_name_to_temporary_name[name]+ " = " + to_param_address(name) + " /* " + name + "*/;\n")
```

Comment:
It seems quite strange for me at the end in the *_Dyn.cpp in the setFOmc to have this kind of code:

// -------------- call functions ----------
  double external_call_tmp_1 = data->localData[0]->realVars[1] /* MyModel.b*/;
  double external_call_tmp_2 = data->localData[0]->realVars[2] /* MyModel.c*/;
  double external_call_tmp_0 = data->localData[0]->realVars[0] /* MyModel.a*/;
  double external_call_tmp_3 = data->localData[0]->realVars[3] /* MyModel.d*/;
{
  external_call_tmp_0 /* MyModel.a variable */ = omc_testFunction2( data->localData[0]->realVars[6] /* MyModel.x variable */, data->localData[0]->realVars[7] /* MyModel.y variable */ ,&external_call_tmp_1 /* MyModel.b variable */ ,&external_call_tmp_2 /* MyModel.c variable */ ,&external_call_tmp_3 /* MyModel.d variable */);
}



  // ----- external_call_4 -----
  f[4] = data->localData[0]->realVars[1] /*  MyModel.b */ - ( external_call_tmp_1 );
  // ----- external_call_5 -----
  f[5] = data->localData[0]->realVars[2] /*  MyModel.c */ - ( external_call_tmp_2 );
  // ----- external_call_6 -----
  f[6] = data->localData[0]->realVars[0] /*  MyModel.a */ - ( external_call_tmp_0 );
  // ----- external_call_7 -----
  f[7] = data->localData[0]->realVars[3] /*  MyModel.d */ - ( external_call_tmp_3 );

Wouldn't it be possible to condensate the beginning and not have this line double external_call_tmp_1 = data->localData... and then external_call_tmp_0 = omc_testFunction but directly double external_call_tmp_1 = omc_testFunction for example?

Reply:
It was done this way as  a variable might be an output value of several functions, and thus we want it to take the same value the variable would have taken.
A = f(B,&C) 
D = g(E, &A)